### PR TITLE
add ctz to wasm

### PR DESCRIPTION
**Pull request by MikeHolman**
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/613)

### CODE REVIEW

File: test/wasm/math.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+let ffi = {};
+let wasm = readbuffer("math.wasm");
```
Comment (commonlisp) [inline test/wasm/math.js]: Curious, where is readbuffer() coming from?

Comment (MikeHolman) [inline test/wasm/math.js]: @Cellule added this as closer analog to how browser behaves (and it is emit by emscripten)

**Comment by Cellule** [comment]
Reviewed 23 of 26 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/Backend/LowerMDShared.cpp, line 1469 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDd_xqxzNuJyt9dsQSL:-KDd_xqxzNuJyt9dsQSM:-1333035854) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Backend/LowerMDShared.cpp#L1469)):*
Is this a change for Wasm ? it looks like a bug fix.

---

*[lib/Runtime/ByteCode/OpCodes.h, line 677 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdZgLhyqLFKrZ-8DX5:-KDdZgLhyqLFKrZ-8DX6:-1257149170) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Runtime/ByteCode/OpCodes.h#L677)):*
The location for the opcode seems odd. Is it a Math Built-in ? If yes name it `InlineMathCtz32` or simply `MathCtz32` if the inlining doesn't make sense. 
Otherwise why not call it `Ctz_I4` and put it closer to `Add_I4`.

---

*[lib/Runtime/Math/WasmMath.h, line 13 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdYup3A7psS7eE9NBg:-KDdYup4IontChgqQGvv:-1733317540) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Runtime/Math/WasmMath.h#L13)):*
Did you put it there because there is no chance asm.js use this operator ?

---

*[test/wasm/math.js, line 9 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdYQds_tIzItFfXAfj:-KDdYQds_tIzItFfXAfk:1113541308) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/test/wasm/math.js#L9)):*
Since there is no baseline, could you add validation in javascript so our test runner will work

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/613)*

**Comment by MikeHolman** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/Backend/LowerMDShared.cpp, line 1469 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDd_xqxzNuJyt9dsQSL:-KDedTLHF8Ixo9WflIVC:-1406214883) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Backend/LowerMDShared.cpp#L1469)):*
Yup. Others weren't calling Legalize and were manually doing it. For my own sanity, I added these to the ops that we can programmatically legalize. 

---

*[lib/Runtime/ByteCode/OpCodes.h, line 677 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdZgLhyqLFKrZ-8DX5:-KDecSmyWRws1TJ6U2sa:-1900602691) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Runtime/ByteCode/OpCodes.h#L677)):*
I did this on purpose. The _I4, _A, etc. in many opcodes is something that I'd like to avoid when possible (and having talked with Louis on this subject, I think he would also prefer to have the opnds determine the type, not the opcode). Would rather not have to come back and create an _I8 for this. It is not a math builtin. There are no math builtins. It is an opcode, and not distinguished in any way from add, sub, etc., so calling it inline builtin would be misleading.

---

*[lib/Runtime/Math/WasmMath.h, line 13 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdYup3A7psS7eE9NBg:-KDedBFEzEGWkXmEoQ5_:-668634129) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/lib/Runtime/Math/WasmMath.h#L13)):*
Yes. I'd like to start distinguishing between things that are wsam only operations, and not intermix them with the ops that are shared with asm.js. At some point I want to make it more explicit what is asm only, what is wasm only, and what is shared. But for now this is a start.

---

*[test/wasm/math.js, line 9 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/613#-KDdYQds_tIzItFfXAfj:-KDedQzCXEVnMMkkYZ5K:-358972885) ([raw file](https://github.com/microsoft/chakracore/blob/5697f46fb7b950981f0c119820977e3c5ccdbf26/test/wasm/math.js#L9)):*
Sure, will do

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/613)*

**Comment by MikeHolman** [comment]
@dotnet-bot test this

**Comment by Cellule** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/613#-:-KE8Nj7HFLwPoFsmsc3s:1417088721)*

**Comment by MikeHolman** [comment]
@dotnet-bot test this

**Pull request merged**